include/net/xia_ether.h

```diff
@@ -240,6 +243,105 @@ static int attach_neigh_addr_to_fib_entry(struct fib_xid_ether_main
 	return 0;
 }
 
+/* Template struct containing hdr manipulation funcs. */
+struct xia_header_ops {
+	/* Creates a new L2 header. */
+	int (*create) (struct sk_buff *skb, struct net_device *dev,
+			   unsigned short type, const void *daddr,
+			   const void *saddr, unsigned int len);
+
+	/* Extracts hrdw address from pkt. */
+	int (*parse)(const struct sk_buff *skb, unsigned char *haddr);
+
+	/* Creates a cached hdr and alocates to a mfxid. */
+	int (*cache)(const struct fib_xid_ether_main *mfxid, struct hh_cache *hh, __be16 type);
+
+	/* Updates the cached hdr on change of address
+	 * type : 1 -> destionation addr
+	 *        0 -> source addr
+	 */
+	void (*cache_update)(struct hh_cache *hh, const struct net_device *dev,
+				const unsigned char *haddr, const int type);
+};
+
+int xia_ether_header_cache(const struct fib_xid_ether_main *mfxid, 
+        struct hh_cache *hh, __be16 type)
+{
+	struct ethhdr *eth;
+	const struct net_device *dev = mfxid->host_interface;
+
+	eth = (struct ethhdr *)
+	    (((u8 *) hh->hh_data) + (HH_DATA_OFF(sizeof(*eth))));
+
+	if (type != htons(ETH_P_XIP))
+		return -1;
+
+	eth->h_proto = type;
+	memcpy(eth->h_source, dev->dev_addr, ETH_ALEN);
+	memcpy(eth->h_dest, mfxid->neigh_addr->ha, ETH_ALEN);
+	hh->hh_len = ETH_HLEN;
+	return 0;
+}
+
+void xia_ether_header_cache_update(struct hh_cache *hh,
+			     const struct net_device *dev,
+			     const unsigned char *haddr,
+			     const int type)
+{
+	struct ethhdr *eth;
+	eth = (struct ethhdr *)
+		    (((u8 *) hh->hh_data) + (HH_DATA_OFF(sizeof(*eth))));
+	if(type)
+		memcpy(eth->h_dest , haddr, ETH_ALEN);
+	else
+		memcpy(eth->h_source, dev->dev_addr, ETH_ALEN);
+}
+
+const struct xia_header_ops xia_ether_hdr_ops ____cacheline_aligned = {
```

Comment:
This is the only place where `struct xia_header_ops` is used. In addition, the fields `create` and `parse` of variable `xia_ether_hdr_ops` are never used. Thus, we don't need variable `xia_ether_hdr_ops`, we can just call `xia_ether_header_cache()` and `xia_ether_header_cache_update()` where needed.